x/nft/username/handler.go

```diff
@@ -249,3 +249,71 @@ func findActor(h tokenHandler, ctx weave.Context, t Token) weave.Address {
 	}
 	return nil
 }
+
+type AddRemoveApprovalHandler struct {
+	tokenHandler
+}
+
+func (h *AddRemoveApprovalHandler) Check(ctx weave.Context, store weave.KVStore, tx weave.Tx) (weave.CheckResult, error) {
+	var res weave.CheckResult
+	if _, err := h.validate(ctx, tx); err != nil {
+		return res, err
+	}
+	res.GasAllocated += createUsernameCost
+	return res, nil
+}
+
+func (h *AddRemoveApprovalHandler) Deliver(ctx weave.Context, store weave.KVStore, tx weave.Tx) (weave.DeliverResult, error) {
+	var res weave.DeliverResult
+	msg, err := h.validate(ctx, tx)
+	if err != nil {
+		return res, err
+	}
+
+	o, t, err := loadToken(h.tokenHandler, store, msg.GetId())
+	if err != nil {
+		return res, err
+	}
+
+	actor := findActor(h.tokenHandler, ctx, t)
+	if actor == nil {
+		return res, errors.ErrUnauthorized()
+	}
+
+	switch v := interface{}(msg).(type) {
+	case *AddApprovalMsg:
+		base := v.Base
+
+		err = t.Approvals().Grant(base.Action, base.Address, base.Options, 0)
+		if err != nil {
+			return res, err
+		}
+	case *RemoveApprovalMsg:
+		base := v.Base
+
+		err = t.Approvals().Revoke(base.Action, base.Address)
+		if err != nil {
+			return res, err
+		}
+	}
+
+	return res, h.bucket.Save(store, o)
+}
+
+func (h *AddRemoveApprovalHandler) validate(ctx weave.Context, tx weave.Tx) (nft.Identified, error) {
+	msg, err := tx.GetMsg()
+	if err != nil {
+		return nil, err
+	}
+
+	imsg := interface{}(msg)
```

Comment:
q: why do you need this cast? msg is already an interface

Reply:
Cause you can't type switch over something that is not interface{}, and it's weave.Msg, afair.

Reply:
weave.Msg is an interface already. Please remove this line.